src/state_machine.cpp

```diff
@@ -0,0 +1,254 @@
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/printk.h>
+#include <inttypes.h>
+#include <zephyr/smf.h>
+#include <zephyr/logging/log.h>
+#include <led_handler.h>
+
+LOG_MODULE_REGISTER(state_machine, LOG_LEVEL_DBG);
+
+#define THREAD_STATE_SIZE 1028 // arbitrary for now
+
+#define DATA_ENABLE DT_ALIAS(databutton) // Button for initiating or cancelling test P1.00
+#if !DT_NODE_HAS_STATUS(DATA_ENABLE, okay)
+#error "Unsupported board: DATA_ENABLE devicetree alias is not defined"
+#endif
+
+/* List of events */
+#define EVENT_BTN_PRESS BIT(0)
+
+static const struct gpio_dt_spec data_enable_button = GPIO_DT_SPEC_GET_OR(DATA_ENABLE, gpios, {0});
+static struct gpio_callback button_cb_data;
+
+/* Forward declaration of state table */
+extern const struct smf_state dev_states[];
+
+enum dev_state { INIT, IDLE, TEST, PROCESS, COMPLETE, CANCEL };
+
+/* User defined object */
+struct s_object {
+    struct smf_ctx ctx;
+
+    /* Events */
+    struct k_event button_press_event;
+    int32_t events;
+
+} s_obj;
```

Comment:
Like in previous comments, I'm a bit hesitant to keep things global. Would it take long to encapsulate this in a static class?